license-gather: add a task for verifying license compatibility

vlsi opened the pull request:
Fixes #46

vlsi commented:
Alternative option:

```kotlin
enum class CompatibilityType {
    ALLOW, UNKNOWN, REJECT;
}

data class LicenseCompatibility(
    val type: CompatibilityType,
    val reason: String
)
```

vlsi commented:
@DreierF , WDYT?

See https://github.com/vlsi/vlsi-release-plugins/pull/49/files#diff-fe8458ed996db9a41bf2a2d964d994e3fe4debc06be8d33ce83c60bb706de5faR97-R118

DreierF commented on plugins/license-gather-plugin/src/test/kotlin/com/github/vlsi/gradle/license/VerifyLicenseCompatibilityTaskTest.kt:
```diff
@@ -0,0 +1,215 @@
+                // No reason, for test purposes
+                allow(LicenseExpressionExtensions.orLater(SpdxLicense.SAX_PD))
+                // https://touk.pl/blog/2018/05/28/testing-kotlin-with-spock-part-2-enum-with-instance-method/
+                AsfLicenseCategory.@A.licenseExpressions.forEach {
+                    allow(it) {
+                        because("The ASF category A is allowed")
+                    }
+                }
```
This overload for `allow` would make things a bit more ergonomic to write. WDYT?

```suggestion
                allow(AsfLicenseCategory.@A) {
                    because("The ASF category A is allowed")
                }
```

vlsi replied:
What I do not like is 6 overloads for allow-reject-unknown and license-licenseExpresion.

Adding category adds 3 more, and it probaby makea sense to add LicenseCategory interface to avoid adding more and more overloads

vlsi replied:
I made `License: LicenseExpressionSet` and `LicenseExpression: LicenseExpressionSet`, and `AsfLicenseCategory: LicenseExpressionSet`

DreierF commented on plugins/license-gather-plugin/src/test/kotlin/com/github/vlsi/gradle/license/VerifyLicenseCompatibilityTaskTest.kt:
```diff
@@ -0,0 +1,215 @@
+                allow(SpdxLicense.EPL_2_0) {
+                    because("JUnit is OK")
+                }
+                allow(new SimpleLicense("The W3C License", uri("http://www.w3.org/TR/2004/REC-DOM-Level-3-Core-20040407/java-binding.zip"))) {
```
I somehow was not able to make this work with licenses that have no URI as for example `org.json:json`.

I tried 
```
allow(new SimpleLicense('JSON', (URI) null))
```
and
```
allow(new SimpleLicense('JSON', List.of()))
```

but both result in 
```
  UNKNOWN: No rules found for JSON
  ================================

  JSON
  * org.json:json:20160212
```

vlsi replied:
Have you tried `SpdxLicense.JSON`?

DreierF replied:
Ah 💡 . I that explains it. I didn't even consider that this specific JSON license or the other license (TMate) for which I bumped into this problem, could be part of SPDX.

vlsi replied:
It is a well-known "non-free" license: https://www.apache.org/legal/resolved.html#category-x, https://wiki.debian.org/qa.debian.org/jsonevil :-/

However, I would add the license normalizer anyway, so `new SimpleLicense("JSON")` would work

vlsi replied:
I've added the normalizer, so all `allow(new SimpleLicense("JSON"))`, `allow(new SimpleLicense("JSON", "http://json..."))`, and `allow(SpdxLicense.JSON)` should work.

DreierF reviewed (commented):
Thanks a lot for implementing that suggestion @vlsi!

vlsi commented on plugins/license-gather-plugin/src/main/kotlin/com/github/vlsi/gradle/license/GatherLicenseTask.kt:
```diff
@@ -504,12 +509,12 @@ open class GatherLicenseTask @Inject constructor(
-                    logger.debug("Detected license for ${e.key}: $license")
-                    e.setValue(e.value.copy(license = license))
-                }
+                jar.manifest.mainAttributes.getValue("Bundle-License")
```
Do you remember the offending dependency?

DreierF replied:
Seems to be `javax.inject:javax.inject:1`

DreierF commented:
@vlsi Do you have any timeline for when you plan to release this?

vlsi commented:
Oh, thanks for the ping. I somehow forgot about all this.
Just wondering: do you think it is all good? (I don't remember if I intended to make more changes)

DreierF commented:
The functionality LGTM as it is. I also have a working version running against a bigger project.

DreierF commented:
@vlsi Is there anything missing from your point of view?

vlsi commented:
@DreierF , I've released 1.78, so please let me know if it works for you.

DreierF commented:
First of all thanks for your work on this!

All in all it works as expected and I'm glad that I can start using it :)
However, I found some minor issues regarding Groovy compatibility:

```groovy
	allow(AsfLicenseCategory.A) {
		because("The ASF category A is allowed")
	}
```
does not seem to work.
It results in 
```
Could not find method allow() for arguments [class com.github.vlsi.gradle.release.AsfLicenseCategory$A, precompiled_ProjectLicense$_run_closure2$_closure11@443721af] on task ':verifyLicense' of type com.github.vlsi.gradle.license.VerifyLicenseCompatibilityTask.
```

Another minor issue is the usage of the `and` and `or` operators. Not sure how this was intended to work. What works is:
```groovy
import static com.github.vlsi.gradle.license.api.LicenseExpressionExtensions.and
import static com.github.vlsi.gradle.license.api.LicenseExpressionExtensions.or

...
	overrideLicense("...") {
		expectedLicense = and(and(Apache_2_0, LGPL_2_1), MPL_1_1)
		effectiveLicense = or(or(Apache_2_0, LGPL_2_1_only), MPL_1_1)
	}
```

When the methods would be actual instance methods instead of extension methods this could be written as 
```groovy
	overrideLicense("...") {
		expectedLicense = Apache_2_0 & LGPL_2_1 & MPL_1_1
		effectiveLicense = Apache_2_0 | LGPL_2_1_only | MPL_1_1
	}
```
but I'm not sure if I might be missing some implementation details that would make this very hard to achieve.
https://groovy-lang.org/operators.html#Operator-Overloading

vlsi commented:
>does not seem to work

Does it happen with 1.78?
I think the test uses allow(....Category.A), it uses groovy DSL, and the test seems to pass

>When the methods would be actual instance methods

The methods should be instance in 1.78.
Do they work for you? 

DreierF commented:
🙈 I used 1.78, but completely forgot that I still had an intermediate version in mavenLocal that was picked up instead of the final 1.78 from Maven Central. That resolved all mentioned issues.
Sorry for the false alarm.

vlsi commented:
No worries 😉 . I added the mentioned changes just a couple of days ago.

My initial intention was to keep `License` interface minimal, however, now I see the conversion to `LicenseExpression` is almost always needed, so I just gave up with the said extensions and moved them to the instance methods.
